NullPointerException in NvdCveClient (if Endpoint is not available?)

It seems, that under some circumstances if the endpoint is not available or does not work as epxected an NPE is thrown.

Client: v5.1.0, 
Gradle Plugin: 9.0.4

https://github.com/jeremylong/Open-Vulnerability-Project/blob/v5.1.0/open-vulnerability-clients/src/main/java/io/github/jeremylong/openvulnerability/client/nvd/NvdCveClient.java#L322
```
> Task :dependencyCheckAnalyze
Verifying dependencies for project emailJava
Checking for updates and analyzing dependencies for vulnerabilities
Error updating the NVD Data
org.owasp.dependencycheck.data.update.exception.UpdateException: Error updating the NVD Data
	at org.owasp.dependencycheck.data.update.NvdApiDataSource.processApi(NvdApiDataSource.java:345)
	at org.owasp.dependencycheck.data.update.NvdApiDataSource.update(NvdApiDataSource.java:110)
	at org.owasp.dependencycheck.Engine.doUpdates(Engine.java:906)
	at org.owasp.dependencycheck.Engine.initializeAndUpdateDatabase(Engine.java:711)
	at org.owasp.dependencycheck.Engine.analyzeDependencies(Engine.java:637)
	at org.owasp.dependencycheck.gradle.tasks.AbstractAnalyze.analyze(AbstractAnalyze.groovy:100)
	at java.base@11.0.21/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.21/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.21/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.21/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:125)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:58)
...
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:380)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.AbstractManagedExecutor$1.run(AbstractManagedExecutor.java:47)
	at java.base@11.0.21/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base@11.0.21/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base@11.0.21/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at io.github.jeremylong.openvulnerability.client.nvd.NvdCveClient.next(NvdCveClient.java:322)
	at io.github.jeremylong.openvulnerability.client.nvd.NvdCveClient.next(NvdCveClient.java:336)
	at org.owasp.dependencycheck.data.update.NvdApiDataSource.processApi(NvdApiDataSource.java:324)
	... 127 more
Unable to update 1 or more Cached Web DataSource, using local data instead. Results may not include recent vulnerabilities.Unable to continue dependency-check analysis.
Region [NODEAUDIT] : Not alive and dispose was called, filename: NODEAUDIT
Region [CENTRAL] : Not alive and dispose was called, filename: CENTRAL
Region [POM] : Not alive and dispose was called, filename: POM
> Task :dependencyCheckAnalyze FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':dependencyCheckAnalyze'.
> Analysis failed.
```

--- a/buildSrc/src/main/groovy/vuln.tools.java-common-conventions.gradle
+++ b/buildSrc/src/main/groovy/vuln.tools.java-common-conventions.gradle
@@ -12,7 +12,7 @@ plugins {
 }
 
 group 'io.github.jeremylong'
-version = '5.1.0'
+version = '5.1.1'
 
 repositories {
     mavenCentral()

--- a/open-vulnerability-clients/README.md
+++ b/open-vulnerability-clients/README.md
@@ -39,14 +39,14 @@ See API usage examples in the [open-vulnerability-store](https://github.com/jere
 <dependency>
    <groupId>io.github.jeremylong</groupId>
    <artifactId>open-vulnerability-clients</artifactId>
-   <version>5.1.0</version>
+   <version>5.1.1</version>
 </dependency>
 ```
 
 ### gradle
 
 ```groovy
-implementation 'io.github.jeremylong:open-vulnerability-clients:5.1.0'
+implementation 'io.github.jeremylong:open-vulnerability-clients:5.1.1'
 ```
 
 ### api usage

--- a/open-vulnerability-clients/src/main/java/io/github/jeremylong/openvulnerability/client/nvd/NvdCveClient.java
+++ b/open-vulnerability-clients/src/main/java/io/github/jeremylong/openvulnerability/client/nvd/NvdCveClient.java
@@ -319,37 +319,43 @@ public class NvdCveClient implements PagedDataSource<DefCveItem> {
         RateLimitedCall call;
         try {
             call = getCompletedFuture();
-            SimpleHttpResponse response = call.getResponse();
-            if (response.getCode() == 200) {
-                LOG.debug("Content-Type Received: {}", response.getContentType());
-                json = response.getBodyText();
-                // resolve issue #20
-                if (json == null && response.getBody().isBytes()) {
-                    json = new String(response.getBodyBytes(), StandardCharsets.UTF_8);
-                }
-
-                CveApiJson20 current;
-                try {
-                    current = objectMapper.readValue(json, CveApiJson20.class);
-                    this.indexesToRetrieve.remove(call.getStartIndex());
-                } catch (JsonProcessingException e) {
+            if (call == null) {
+                if (hasNext()) {
                     return next();
                 }
-                this.totalAvailable = current.getTotalResults();
-                lastUpdated = findLastUpdated(lastUpdated, current.getVulnerabilities());
-                if (firstCall) {
-                    firstCall = false;
-                    queueCalls();
-                }
-                if (futures.isEmpty() && !indexesToRetrieve.isEmpty()) {
-                    queueUnsuccessful();
-                }
-                return current.getVulnerabilities();
             } else {
-                lastStatusCode = response.getCode();
-                LOG.debug("Status Code: {}", lastStatusCode);
-                LOG.debug("Response: {}", response.getBodyText());
-                throw new NvdApiException("NVD Returned Status Code: " + lastStatusCode);
+                SimpleHttpResponse response = call.getResponse();
+                if (response.getCode() == 200) {
+                    LOG.debug("Content-Type Received: {}", response.getContentType());
+                    json = response.getBodyText();
+                    // resolve issue #20
+                    if (json == null && response.getBody().isBytes()) {
+                        json = new String(response.getBodyBytes(), StandardCharsets.UTF_8);
+                    }
+
+                    CveApiJson20 current;
+                    try {
+                        current = objectMapper.readValue(json, CveApiJson20.class);
+                        this.indexesToRetrieve.remove(call.getStartIndex());
+                    } catch (JsonProcessingException e) {
+                        return next();
+                    }
+                    this.totalAvailable = current.getTotalResults();
+                    lastUpdated = findLastUpdated(lastUpdated, current.getVulnerabilities());
+                    if (firstCall) {
+                        firstCall = false;
+                        queueCalls();
+                    }
+                    if (futures.isEmpty() && !indexesToRetrieve.isEmpty()) {
+                        queueUnsuccessful();
+                    }
+                    return current.getVulnerabilities();
+                } else {
+                    lastStatusCode = response.getCode();
+                    LOG.debug("Status Code: {}", lastStatusCode);
+                    LOG.debug("Response: {}", response.getBodyText());
+                    throw new NvdApiException("NVD Returned Status Code: " + lastStatusCode);
+                }
             }
         } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
@@ -362,8 +368,8 @@ public class NvdCveClient implements PagedDataSource<DefCveItem> {
                 return next();
             }
             close();
-            return null;
         }
+        return null;
     }
 
     /**
@@ -389,17 +395,17 @@ public class NvdCveClient implements PagedDataSource<DefCveItem> {
     }
 
     private RateLimitedCall getCompletedFuture() throws InterruptedException, ExecutionException {
-        boolean notFound = futures.size() > 0;
         Future<RateLimitedCall> result = null;
-        while (notFound) {
+        while (result == null && !futures.isEmpty()) {
             for (Future<RateLimitedCall> future : futures) {
                 if (future.isDone()) {
                     result = future;
-                    notFound = false;
                     break;
                 }
             }
-            Thread.sleep(500);
+            if (result == null) {
+                Thread.sleep(500);
+            }
         }
         if (result != null) {
             futures.remove(result);

--- a/vulnz/README.md
+++ b/vulnz/README.md
@@ -72,7 +72,7 @@ export JAVA_OPTS="-Xmx2g"
 Alternatively, run the CLI using the `-Xmx2g` argument:
 
 ```bash
-java -Xmx2g -jar ./vulnz-5.1.0.jar
+java -Xmx2g -jar ./vulnz-5.1.1.jar
 ```
 
 ### Creating the Cache
@@ -89,7 +89,7 @@ for file in *.json; do gzip -k "${file}"; done
 Alternatively, without using the above install command:
 
 ```bash
-./vulnz-5.1.0.jar cve --cache --directory ./cache
+./vulnz-5.1.1.jar cve --cache --directory ./cache
 cd cache
 for file in *.json; do gzip -k "${file}"; done 
 ```

